
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
  </head>
<body ng-app="app">
<div class="container">
        <h2>Metascape Gene List Analysis Report</h2><p/>  <a href="http://metascape.org">metascape.org</a><sup>1</sup><p/>

      <h3>Gene Lists</h3>
      User-provided gene identifiers are first converted into their corresponding human Entrez gene IDs using the latest version of the database (last updated on 2024-01-101).  If multiple identifiers correspond to the same Entrez gene ID, they will be considered as a single Entrez gene ID in downstream analyses.   The gene lists are summarized in Table 1.<p/>
          <div class="panel panel-info">
            <div class="panel-heading">Table 1. Statistics of input gene lists.</div>
            <div class="panel-body"><TABLE class="table">
  <THEAD>
    <TR>
      <TH class="info">Name</TH>
      <TH class="info">Unique</TH>
    </TR>
  </THEAD>
  <TBODY>
    <TR>
      <TD>Trpc3</TD>
      <TD>6</TD>
    </TR>
  </TBODY>
</TABLE></div>
          </div>




<h3>Protein-protein Interaction Enrichment Analysis</h3>
For each given gene list, protein-protein interaction enrichment analysis has been carried out with the following databases: STRING<sup>2</sup>, BioGrid<sup>3</sup>, OmniPath<sup>4</sup>, InWeb_IM<sup>5</sup>.Only physical interactions in STRING (physical score &gt 0.132) and BioGrid are used (<a href="http://metascape.org/blog/?p=219">details</a>).  The resultant network contains the subset of proteins that form physical interactions with at least one other member in the list.  If the network contains between 3 and 500 proteins, the Molecular Complex Detection (MCODE) algorithm<sup>6</sup> has been applied to identify densely connected network components.  

<h3>Quality Control and Association Analysis</h3>

<p>Gene list enrichments are identified in the following ontology categories: Cell_Type_Signatures. All genes in the genome have been used as the enrichment background. Terms with a p-value &lt; 0.01, a minimum count of 3, and an enrichment factor &gt; 1.5 (the enrichment factor is the ratio between the observed counts and the counts expected by chance) are collected and grouped into clusters based on their membership similarities.  The top few enriched clusters (one term per cluster) are shown in the Figure 1.  The algorithm used here is the same as that is used for pathway and process enrichment analysis.<p/>

              <div class="panel panel-info">
                <div class="panel-heading">Figure 1. Summary of enrichment analysis in Cell Type Signatures<sup>7</sup>.</div>
                <div class="panel-body">

                <table>
                    <tr>
                        <td>
                            <img src="./Enrichment_QC/HeatmapSelectedGO_Cell_Type_Signatures.png" style="width:1000px;">
                        </td>
                    </tr>
                    <tr>
                        <td align='center'>
                            
                <a href='./Enrichment_QC/HeatmapSelectedGO_Cell_Type_Signatures.pdf' title='download PDF file'>
                <img class='link' src='icon/PDF48.png' >
                </a>
                
                        </td>
                    </tr>
                    <tr>
                        <td>
                            <br><br><TABLE class="table">
  <THEAD>
    <TR>
      <TH class="info">GO</TH>
      <TH class="info">Description</TH>
      <TH class="info">Count</TH>
      <TH class="info">%</TH>
      <TH class="info">Log10(P)</TH>
      <TH class="info">Log10(q)</TH>
    </TR>
  </THEAD>
  <TBODY>
    <TR>
      <TD><a href="https://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?systematicName=M39245" target="_TERM">M39245</a></TD>
      <TD>LAKE ADULT KIDNEY C26 MESANGIAL CELLS</TD>
      <TD>3</TD>
      <TD>50</TD>
      <TD>-5.40</TD>
      <TD>-0.91</TD>
    </TR>
    <TR>
      <TD><a href="https://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?systematicName=M39040" target="_TERM">M39040</a></TD>
      <TD>FAN EMBRYONIC CTX BRAIN ENDOTHELIAL 2</TD>
      <TD>3</TD>
      <TD>50</TD>
      <TD>-4.70</TD>
      <TD>-0.51</TD>
    </TR>
    <TR>
      <TD><a href="https://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?systematicName=M39018" target="_TERM">M39018</a></TD>
      <TD>FAN EMBRYONIC CTX BIG GROUPS BRAIN ENDOTHELIAL</TD>
      <TD>3</TD>
      <TD>50</TD>
      <TD>-4.50</TD>
      <TD>-0.44</TD>
    </TR>
    <TR>
      <TD><a href="https://www.gsea-msigdb.org/gsea/msigdb/geneset_page.jsp?systematicName=M39175" target="_TERM">M39175</a></TD>
      <TD>MURARO PANCREAS MESENCHYMAL STROMAL CELL</TD>
      <TD>3</TD>
      <TD>50</TD>
      <TD>-3.70</TD>
      <TD>0.00</TD>
    </TR>
  </TBODY>
</TABLE>
                        </td>
                    </tr>
                </table>

                </div>
              </div>
    <h3>Reference</h3></p>
    <ol style="list-style: decimal inside;">
      <li>Zhou et al., Metascape provides a biologist-oriented resource for the analysis of systems-level datasets. Nature Communications (2019) 10(1):1523.</li>
      <li>Szklarczyk D. et al. STRING v11: protein-protein association networks with increased coverage, supporting functional discovery in genome-wide experimental datasets. Nucleic Acids Res. (2019) 47:D607-613.</li>
      <li>Stark C. et al. BioGRID: a general repository for interaction datasets. Nucleic Acids Res. (2006) 34:D535-539.</li>
      <li>Turei D. et al. A scored human protein-protein interaction network to catalyze genomic interpretation. Nat. Methods. (2016) 13:966-967.</li>
      <li>Li T. et al. A scored human protein-protein interaction network to catalyze genomic interpretation. Nat. Methods. (2017) 14:61-64.</li>
      <li>Bader, G.D. et al. An automated method for finding molecular complexes in large protein interaction networks. BMC bioinformatics (2003) 4:2.</li>
      <li>Subramanian A, et al. Gene set enrichment analysis: A knowledge-based approach for interpreting genome-wide expression profiles. Proc Natl Acad Sci U S A 102, 15545-15550 (2005).</li>
</ol>
</div>

</body></html>