---
title: "10X QC processing"
author: "Kimberly Olney, PhD"
date: "05/08/2024"
output:
  html_document:
    theme: cerulean
    toc: true
    toc_float: true
editor_options: 
  chunk_output_type: inline
---
# Post-mortem interval (PMI)
- Fresh versus 3 hours 
- Single nucleus RNAseq 
- Male and female mice 

# Set working directory
```{r working_directory}
knitr::opts_knit$set(root.dir = ".")
```

# Libraris, paths, colors
```{r echo=FALSE, message=FALSE}
source(here::here("/research/labs/neurology/fryer/m239830/PMI/scripts", "file_paths_and_colours.R"))
treatment <- "PMI"
samples <- c("P_69", "B_7", "P_65", "B_6", "P_52", "P_62", "P_67", "J_3", "J_4", "P_60", "P_71", "B_10", "B_8", "B_9", "P_68", "P_54", "P_53", "P_64", "J_2", "J_1")
time_order <- c("0hr", "0hr", "0hr", "0hr", "0hr", "0hr", "0hr", "0hr", "0hr", "3hr", "3hr", "3hr", "3hr", "3hr", "3hr", "3hr", "3hr", "3hr", "3hr", "3hr")
sex_order <- c("Female", "Female", "Female", "Female", "Male", "Male", "Male", "Male", "Male", "Female", "Female", "Female", "Female", "Female", "Male", "Male", "Male", "Male", "Male", "Male")
strain_order <- c("PS19", "B6", "PS19", "B6", "PS19", "PS19", "PS19", "JAX", "JAX", "PS19", "PS19", "B6", "B6", "B6", "PS19", "PS19", "PS19", "PS19", "JAX", "JAX")
group_order <- paste0(time_order, ".", sex_order, ".", strain_order)
```

# Genes 
```{r gene_info}
if (file.exists("../rObjects/annotation.rds")) {
  genes <- readRDS("../rObjects/annotation.rds")
} else {
  gtf.file <- paste0(pathToRef, "/genes/genes.gtf")
  genes <- rtracklayer::import(gtf.file)
  genes <- as.data.frame(genes)
  genes <- genes[genes$type == "gene",]
  saveRDS(genes, "../rObjects/annotation.rds")
}

gene_type_table <- table(genes$gene_type)
write.table(gene_type_table, "gene_type_table.tsv", row.names = F, quote = F, sep = "\t")
mt.genes.df <- subset(genes, seqnames == "chrM")
mt.genes <- mt.genes.df$gene_name
```

# QC thresholds
nCount_RNA = total number of transcripts (UMIs) in a single cell 
nFeature_RNA = number of unique genes (features)
```{r QC_metrics}
nCount.min <- 500 
nCount.max <- 100000 
nFeature.min <- 100
complexity.cutoff <- 0.80
mt.cutoff <- 4 
hb.cutoff <- 1
ribo.cutoff <- 5
```

# Setup seurat object
```{r seurat_object}
prefix <- "../cellranger/"
suffix <- "/outs/filtered_feature_bc_matrix.h5"

if (file.exists(paste0("../rObjects/", treatment, ".rds"))) {
  dataObject <- readRDS(paste0("../rObjects/", treatment, ".rds"))
} else {
  for (i in samples) {
    print(i)
    obj <- CreateSeuratObject(Read10X_h5(paste0(prefix,i,suffix)))
    assign(i, obj)
  }
  # merge objects
  dataObject <- merge(x = P_69,
                 y = c(B_7, P_65, B_6, P_52, P_62, P_67, J_3, J_4, P_60, P_71, 
                       B_10, B_8, B_9, P_68, P_54, P_53, P_64, J_2, J_1),
                 add.cell.ids = samples, 
                 project = "PMI")
    saveRDS(dataObject, paste0("../rObjects/", treatment, ".rds"))
} 
# Inspect 
dataObject
```

# Change indents
```{r change_idents}
barcodes <- colnames(dataObject)
sample <- str_match(barcodes, "(.+)_[ACGT]+")[,2]
table(sample)
dataObject$sample <- factor(sample, levels = samples)
table(dataObject$sample)  # check
Idents(dataObject) <- dataObject$sample
```

Add condition column to metadata
```{r condition_column}
# Match the samples to the dataObject
sample_match <- match(dataObject$sample, samples)

# Time
dataObject$time <- time_order[sample_match]
table(dataObject$time) 

# Sex 
dataObject$sex <- sex_order[sample_match]
table(dataObject$sex) 

# Strain 
dataObject$strain <- strain_order[sample_match]
table(dataObject$strain) 


# Group
dataObject$group <- group_order[sample_match]
table(dataObject$group) 
```

QC columns 
```{r QC_columns}
summary(dataObject$nCount_RNA)
summary(dataObject$nFeature_RNA)
# cell.complexity
dataObject$cell.complexity <- log10(dataObject$nFeature_RNA) / log10(dataObject$nCount_RNA)

# Chromosome M
gene.names <- rownames(dataObject)
dataObject$percent.mt <- PercentageFeatureSet(dataObject, features = mt.genes)
summary(dataObject$percent.mt)

# ribosomal proteins 
ribo.genes <- gene.names[grep("^Rp[sl]", gene.names)] 
mt.ribo <- gene.names[grep("^Mrp[sl]", gene.names)]
ribo.combined <- c(mt.ribo,ribo.genes)
dataObject$percent.ribo <- PercentageFeatureSet(dataObject, features = ribo.combined)
summary(dataObject$percent.ribo)

# hemoglobin proteins
hb.genes <- gene.names[grep("^Hb[ba]-", gene.names)]
dataObject$percent.hb <- PercentageFeatureSet(dataObject, features = hb.genes)
summary(dataObject$percent.hb)

# percent choroid plexus
dataObject$percent.choroid <- PercentageFeatureSet(dataObject, features = c("Ttr","Folr1", "Prlr"))
summary(dataObject$percent.choroid)
```

# Pre-filtering QC
## Number of cells
```{r prefiltering_cells_per_sample}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject$sample))
colnames(data) <- c("sample","frequency")

ncells1 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,20000, by = 5000), limits = c(0,20000)) +
  ggtitle("Raw: nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))
ncells1
```

## Density plots
```{r prefiltering_density}
# set graphical parameter
par(mfrow = c(3,1))

# Visualize nCount_RNA
den1 <- ggplot(dataObject@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max) 

# Visualize percent.mt
den2 <- ggplot(dataObject@meta.data,
       aes(color = sample,
           x = percent.mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_continuous(n.breaks = 4) +
  geom_vline(xintercept = mt.cutoff) +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density")

# Visualize cell complexity
# Quality cells are usually above 0.85
den3 <- ggplot(dataObject@meta.data,
       aes(color = sample,
           x = cell.complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff)

# Arrange graphs in grid
plots1 <- list(den1,den2,den3)
layout1 <- rbind(c(1),c(2),c(3))
grid1 <- grid.arrange(grobs = plots1, layout_matrix = layout1)
```

## Violin plots
```{r prefiltering_violins}
# nFeature, nCount, and cell.complexity violins
v1 <- VlnPlot(dataObject,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v1

#  percent violins
v2 <- VlnPlot(dataObject,
              features = c("percent.mt","percent.ribo","percent.hb", "percent.choroid"),
              ncol = 4,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v2
```

```{r, echo=FALSE}
# save v1
v1
path <- paste0("../results/violin/",treatment,"_nFeature_nCount_complexity_raw")
saveToPDF(paste0(path, ".pdf"), width = 20, height = 4)
dev.off()

# save v2
v2
path <- paste0("../results/violin/",treatment,"_percent_raw")
saveToPDF(paste0(path, ".pdf"), width = 25, height = 4)
dev.off()

# cleanup
remove(v1,v2)
```

## Scatter plots
```{r prefiltering_scatter1, warning=FALSE}
s1 <- ggplot(
  dataObject@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  stat_smooth(method=lm) +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_vline(xintercept = nCount.max) + 
  geom_hline(yintercept = nFeature.min) + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s1
```

```{r echo=FALSE}
# save
s1
path <- paste0("../results/scatter/",treatment,"_nFeature_vs_nCount_percentMt_raw")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 10)
dev.off()

# cleanup
remove(s1)
```

```{r prefiltering_scatter2}
s2 <- FeatureScatter(dataObject,
               feature1 = "nCount_RNA",
               feature2 = "percent.mt",
               group.by = 'sample',
               cols = sample_colors,
               shuffle = TRUE)
s2
```

```{r,echo=FALSE}
# save
s2
path <- paste0("../results/scatter/",treatment,"_nCount_vs_percentMT_raw")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# cleanup
remove(s2)
```

# Filtering
## Cell-level filtering
We want to be careful filtering because removing things can easily lead to
misinterpretation.  For example, cells with high percent.mt could actually
just be involved in respiratory processes. \
```{r cell_filtering}
# filter
dataObject.filtered <- subset(dataObject,
                        subset = (nCount_RNA > nCount.min)&
                          (nCount_RNA < nCount.max) & 
                          (nFeature_RNA > nFeature.min) & 
                          (cell.complexity > complexity.cutoff) &
                          (percent.mt < mt.cutoff) & 
                          (percent.hb < hb.cutoff) &
                          (percent.ribo < ribo.cutoff))

# print cells removed
print(paste0(dim(dataObject)[2] - dim(dataObject.filtered)[2]," nuclei removed"))

table(dataObject$sample)
table(dataObject.filtered$sample)
```

## Gene-level filtering
Remove lowly expressed genes.  We will keep genes that have at least 1 count in 10 cells.
```{r gene_filtering}
dataObject.filtered <- JoinLayers(dataObject.filtered)
counts <- GetAssayData(object = dataObject.filtered, layer = "counts")
nonzero <- counts > 0 
keep <- Matrix::rowSums(nonzero) >= 10  
counts.filtered <- counts[keep,]  

dataObject.filtered <- CreateSeuratObject(counts.filtered, 
                                     meta.data = dataObject.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))
```

## Mitochondrial gene filtering
```{r}
# remove mt.genes
counts <- GetAssayData(object = dataObject.filtered, layer = "counts")
keep <- !rownames(counts) %in% mt.genes # false when mt.gene
counts.filtered <- counts[keep,]

# overwrite dataObject.filtered
dataObject.filtered <- CreateSeuratObject(counts.filtered, 
                                    meta.data = dataObject.filtered@meta.data)

# print features removed
print(paste0(dim(counts)[1] - dim(counts.filtered)[1], " features removed"))
```

```{r,echo=FALSE}
saveRDS(dataObject.filtered, paste0("../rObjects/",treatment,"_filtered.rds"))
#dataObject.filtered <- readRDS(paste0("../rObjects/",tolower(tissue),"_filtered.rds"))
```


## Number of cells
```{r number_cells2}
# Visualize the number of cell counts per sample
data <- as.data.frame(table(dataObject.filtered$sample))
colnames(data) <- c("sample","frequency")

ncells2 <- ggplot(data, aes(x = sample, y = frequency, fill = sample)) + 
  geom_col() +
  theme_classic() +
  geom_text(aes(label = frequency), 
            position=position_dodge(width=0.9), 
            vjust=-0.25) +
  scale_fill_manual(values = sample_colors) + 
  scale_y_continuous(breaks = seq(0,20000, by = 5000), limits = c(0,20000)) +
  ggtitle("Filtered: nuclei per sample") +
  theme(legend.position =  "none") + 
  theme(axis.text.x = element_text(angle = 45, hjust=1))

# Arrange graphs in grid
plots2 <- list(ncells1,ncells2)
layout2 <- cbind(c(1),c(2))
grid2 <- grid.arrange(grobs = plots2, layout_matrix = layout2)
```

```{r,echo=FALSE}
# save
grid2 <- grid.arrange(grobs = plots2, layout_matrix = layout2)
path <- paste0("../results/nuclei_count/",treatment, 
               "_cells_per_sample")
saveToPDF(paste0(path, ".pdf"), width = 16, height = 4)

# cleanup
remove(ncells1,ncells2,plots2,layout2,grid2)
```
## Density plots
```{r postfiltering_density}
# set graphical parameter
par(mfrow = c(3,1))

# Visualize the number of counts per cell
den4 <- ggplot(dataObject.filtered@meta.data,
       aes(color = sample,
           x = nCount_RNA,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("nCount_RNA") +
  ylab("Density") +
  geom_vline(xintercept = nCount.min) +
  geom_vline(xintercept = nCount.max)

# Visualize percent.mt
den5 <- ggplot(dataObject.filtered@meta.data,
       aes(color = sample,
           x = percent.mt,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("% Mitochondrial Genes") +
  ylab("Density") +
  geom_vline(xintercept = mt.cutoff)

# Visualize cell complexity
# Quality cells are usually above 0.80
den6 <- ggplot(dataObject.filtered@meta.data,
       aes(color = sample,
           x = cell.complexity,
           fill = sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  scale_x_log10() +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("Cell Complexity (log10(nFeature/nCount))") +
  ylab("Density") +
  geom_vline(xintercept = complexity.cutoff)

# Arrange graphs in grid
plots3 <- list(den1,den2,den3,den4,den5,den6)
layout3 <- rbind(c(1,4),c(2,5),c(3,6))
grid3 <- grid.arrange(grobs = plots3, layout_matrix = layout3)
```

```{r, echo=FALSE}
# save
grid3 <- grid.arrange(grobs = plots3, layout_matrix = layout3)
path <- paste0("../results/density/",treatment, 
               "_density")
saveToPDF(paste0(path, ".pdf"), width = 10, height = 6)

# cleanup
remove(den1,den2,den3,den4,den5,den6,plots3,layout3,grid3)
```

## Violin plots
```{r postfiltering_violins}
# nFeature, nCount, and cell.complexity violins
v3 <- VlnPlot(dataObject.filtered,
              features = c("nFeature_RNA", "nCount_RNA","cell.complexity"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v3

#  percent violins
v4 <- VlnPlot(dataObject.filtered,
              features = c("percent.mt","percent.ribo","percent.hb"),
              ncol = 3,
              group.by = 'sample',
              cols = sample_colors,
              pt.size = 0)
v4
```

```{r, echo=FALSE}
# save
v3
path <- paste0("../results/violin/",treatment, 
               "_nFeature_nCount_complexity_filtered")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 4)

v4
path <- paste0("../results/violin/",treatment, 
               "_percent_filtered")
saveToPDF(paste0(path, ".pdf"), width = 15, height = 4)

# cleanup
remove(v3,v4)
```

## Scatter plots
```{r postfiltering_scatter1}
s3 <- ggplot(
  dataObject.filtered@meta.data,
  aes(x = nCount_RNA, y = nFeature_RNA, color = percent.mt)) + 
  geom_point() + 
  stat_smooth(method=lm) +
	scale_x_log10() +   	
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = nCount.min) + 
  geom_hline(yintercept = nFeature.min) + 
  facet_wrap(~sample) +
  scale_colour_gradient(low = "gray90", high = "black", limits =c(0,100))
s3
```

```{r,echo=FALSE}
# save
s3
path <- paste0("../results/scatter/",treatment,
               "_nFeature_vs_nCount_filtered")
saveToPDF(paste0(path, ".pdf"), width = 7, height = 10)

# cleanup
remove(s3)
```

```{r postfiltering_scatter2}
s4 <- FeatureScatter(dataObject.filtered,
               feature1 = "nCount_RNA",
               feature2 = "percent.mt",
               group.by = 'sample',
               cols = sample_colors,
               shuffle = TRUE)
s4
```

```{r,echo=FALSE}
# save
s4
path <- paste0("../results/scatter/",treatment,
               "_nCount_vs_percentMT_filtered")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# cleanup
remove(s4)
```

## Box plot
```{r boxplot}
# Visualize the distribution of genes detected per cell via boxplot
b1 <- ggplot(dataObject.filtered@meta.data,
       aes(x = sample, 
           y = log10(nFeature_RNA), 
           fill=sample)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust = 0.5, face="bold")) +
  ggtitle("Unique Genes / Cell / Sample") +
  scale_color_manual(values = sample_colors) +
  scale_fill_manual(values = sample_colors) +
  xlab("Sample")
b1
```

```{r,echo=FALSE}
# save
b1
path <- paste0("../results/nuclei_count/",treatment,
               "_nFeature_per_sample")
saveToPDF(paste0(path, ".pdf"), width = 12, height = 4)

# cleanup
remove(b1)
```

# Top transcripts
```{r top_transcripts}
df <- data.frame(row.names = rownames(dataObject.filtered))
df$rsum <- rowSums(x = dataObject.filtered, slot = "counts")
df$gene_name <- rownames(df)
df <- df[order(df$rsum,decreasing = TRUE),]
head(df, 10)
```

```{r,echo=FALSE}
write.table(df, paste0("../results/top_transcripts/", treatment, "_abundant_transcripts.txt"),
            quote = FALSE,
            row.names = FALSE)
```


# SCTransform
Now, we can use the SCTransform method as a more accurate method of normalizing, 
estimating the variance of the raw filtered data, and identifying the most 
variable genes. Variation in sequencing depth (total nCount_RNA per cell) is 
normalized using a regularized negative binomial model.Variance is also 
adjusted based on pooling information across genes with similar abundances. 

Sctransform automatically accounts for cellular sequencing depth by regressing 
out sequencing depth (nUMIs). However, if there are other sources of 
uninteresting variation identified in the data during the exploration steps we 
can also include these. We observed little to no effect due to cell cycle phase 
and so we chose not to regress this out of our data. We observed some effect of 
mitochondrial expression and so we choose to regress this out from the data.

Since we have four samples in our dataset (from two conditions), we want to keep 
them as separate objects and transform them as that is what is required for 
integration. We will first split the cells in seurat.phase object by sample. 

```{r split_object2}
dataObject.filtered[["RNA"]] <- split(dataObject.filtered[["RNA"]], 
                                 f = dataObject.filtered$sample)
```

Now we will use a ‘for loop’ to run the SCTransform() on each sample, and 
regress out mitochondrial expression by specifying in the vars.to.regress 
argument of the SCTransform() function.

Before we run this for loop, we know that the output can generate large R 
objects/variables in terms of memory. If we have a large dataset, then we might 
need to adjust the limit for allowable object sizes within R (Default is 500 * 
1024 ^ 2 = 500 Mb) using the following code:

```{r sct}
# transform
dataObject.filtered <- SCTransform(dataObject.filtered, verbose = FALSE)
```

NOTE: By default, after normalizing, adjusting the variance, and regressing out 
uninteresting sources of variation, SCTransform will rank the genes by residual 
variance and output the 3000 most variant genes. If the dataset has larger cell 
numbers, then it may be beneficial to adjust this parameter higher using the 
variable.features.n argument.

Note, the last line of output specifies “Set default assay to SCT”. We can view 
the different assays that we have stored in our seurat object.

A thread about whether or not regress out batch:
https://github.com/satijalab/issues/3270
It is suggested to not regress out batch, and instead use a data integration method 



## PCA
```{r pca}
# run PCA on the merged object
dataObject.filtered <- RunPCA(object = dataObject.filtered, assay = "SCT")
```

```{r plot_pca}
# Reset idents and levels
DefaultAssay(dataObject.filtered) <- "SCT"
Idents(dataObject.filtered) <- "sample"

# Plot PCA
pca1 <- DimPlot(dataObject.filtered,
                reduction = "pca",
                split.by = "sample",
                group.by = "sample",
                ncol = 4)
pca1

DimPlot(dataObject.filtered,
                reduction = "pca")


```

```{r echo=FALSE}
# save
pca1
path <- paste0("../results/pca/",treatment,
               "_pca")
saveToPDF(paste0(path, ".pdf"), width = 15, height = 4)


DimHeatmap(dataObject.filtered, dims = 1:15, cells = 500, balanced = TRUE)
path <- paste0("../results/pca/",treatment,
               "_pca_genes_dim1to15")
saveToPDF(paste0(path, ".pdf"), width = 8, height = 11)

# cleanup
remove(pca1, PCplot)
```

## Top variable features
Top 20 variable features
```{r variable_genes}
top20 <- dataObject.filtered@assays$SCT@var.features[1:20]
top20
```


# Find significant PCs
To overcome the extensive technical noise in the expression of any single gene 
for scRNA-seq data, Seurat assigns cells to clusters based on their PCA scores 
derived from the expression of the integrated most variable genes, with each PC 
essentially representing a “metagene” that combines information across a 
correlated gene set. Determining how many PCs to include in the clustering step 
is therefore important to ensure that we are capturing the majority of the 
variation, or cell types, present in our dataset.

```{r view_PCs}
# Printing out the most variable genes driving PCs
print(x = dataObject.filtered[["pca"]], 
      dims = 1:10, 
      nfeatures = 10)
```

Quantitative approach to an elbow plot
- The point where the principal components only contribute 5% of standard 
  deviation and the principal components cumulatively contribute 90% of the 
  standard deviation.
- The point where the percent change in variation between the consecutive PCs is 
  less than 0.1%.

First metric
```{r metric1}
# Determine percent of variation associated with each PC
stdv <- dataObject.filtered[["pca"]]@stdev
sum.stdv <- sum(dataObject.filtered[["pca"]]@stdev)
percent.stdv <- (stdv / sum.stdv) * 100

# Calculate cumulative percents for each PC
cumulative <- cumsum(percent.stdv)

# Determine which PC exhibits cumulative percent greater than 90% and
# and % variation associated with the PC as less than 5
co1 <- which(cumulative > 90 & percent.stdv < 5)[1]
co1
```

Second metric
```{r metric2}
# Determine the difference between variation of PC and subsequent PC
co2 <- sort(which(
  (percent.stdv[1:length(percent.stdv) - 1] - 
     percent.stdv[2:length(percent.stdv)]) > 0.1), 
  decreasing = T)[1] + 1

# last point where change of % of variation is more than 0.1%.
co2
```

Choose the minimum of these two metrics as the PCs covering 
the majority of the variation in the data.
```{r minimum_PCs}
# Minimum of the two calculation
min.pc <- min(co1, co2)
min.pc
```

## Elbow plot
Use min.pc we just calculated to generate the clusters. We can plot the elbow 
plot again and overlay the information determined using our metrics:
```{r quantitative_elbow}
# Create a dataframe with values
plot_df <- data.frame(pct = percent.stdv, 
           cumu = cumulative, 
           rank = 1:length(percent.stdv))

# Elbow plot to visualize 
  ggplot(plot_df, aes(cumulative, percent.stdv, label = rank, color = rank > min.pc)) + 
  geom_text() + 
  geom_vline(xintercept = 90, color = "grey") + 
  geom_hline(yintercept = min(percent.stdv[percent.stdv > 5]), color = "grey") +
  theme_bw()
```

# UMAP
```{r UMAP, message=FALSE, warning=FALSE}
# Run UMAP
dataObject.filtered <- RunUMAP(dataObject.filtered,
                           dims = 1:min.pc,
                           reduction = "pca",
                           n.components = 3) # set to 3 to use with VR
```

```{r plot_UMAP}
# plot UMAP and color based on treatment
DimPlot(dataObject.filtered,
        group.by = "sample",
        split.by = "sample",
        shuffle = TRUE)

DimPlot(dataObject.filtered,
        group.by = "sample",
        shuffle = TRUE)
```

# Cluster the cells
Seurat uses a graph-based clustering approach, which embeds cells in a graph 
structure, using a K-nearest neighbor (KNN) graph (by default), with edges drawn 
between cells with similar gene expression patterns. Then, it attempts to 
partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’ 
[Seurat - Guided Clustering Tutorial].

We will use the FindClusters() function to perform the graph-based clustering. 
The resolution is an important argument that sets the “granularity” of the 
downstream clustering and will need to be optimized for every individual experiment. 
For datasets of 3,000 - 5,000 cells, the resolution set between 0.4-1.4 generally 
yields good clustering. Increased resolution values lead to a greater number of 
clusters, which is often required for larger datasets.

The FindClusters() function allows us to enter a series of resolutions and will 
calculate the “granularity” of the clustering. This is very helpful for testing 
which resolution works for moving forward without having to run the function for 
each resolution.

```{r find_neighbors, message=FALSE, warning=FALSE}
# Determine the K-nearest neighbor graph
dataObject.unannotated <- FindNeighbors(object = dataObject.filtered, 
                                 assay = "SCT", # set as default after SCTransform
                                 reduction = "pca", # pca, harmony 
                                 dims = 1:min.pc)

# Determine the clusters for various resolutions
dataObject.unannotated <- FindClusters(object = dataObject.unannotated,
                                 algorithm = 1, # 1= Louvain
                                 resolution = seq(0.1,1.2,by=0.1))
```

```{r}
DimPlot(dataObject.unannotated,
        group.by = "seurat_clusters",
        label = TRUE)
```

# Explore resolutions
```{r}
# 0.1
umap0.1 <- DimPlot(dataObject.unannotated,
        group.by = "SCT_snn_res.0.1",
        label = TRUE)
umap0.1

# 0.4
umap0.4 <- DimPlot(dataObject.unannotated,
        group.by = "SCT_snn_res.0.4",
        label = TRUE)
umap0.4

# 0.6
umap0.6 <- DimPlot(dataObject.unannotated,
        group.by = "SCT_snn_res.0.6",
        label = TRUE)
umap0.6

# 0.8
umap0.8 <- DimPlot(dataObject.unannotated,
        group.by = "SCT_snn_res.0.8",
        label = TRUE)
umap0.8

```


# Clustering QC
## Treatment, sample, phase
```{r umap_treatment}

# sample
u2 <- DimPlot(dataObject.unannotated, 
        label = FALSE,
        group.by = "SCT_snn_res.0.8",
        split.by = "sample") +
  NoLegend()
u2

```


## Revisit QC metrics
```{r umap_QC}
# nCount
f1 <- FeaturePlot(dataObject.unannotated, 
            features = "nCount_RNA",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f1

# nFeature
f2 <- FeaturePlot(dataObject.unannotated, 
            features = "nFeature_RNA",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f2

# percent.mt
f3 <- FeaturePlot(dataObject.unannotated, 
            features = "percent.mt",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f3

# cell.complexity
f4 <- FeaturePlot(dataObject.unannotated, 
            features = "cell.complexity",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f4

```

```{r,echo=FALSE,message=FALSE,warning=FALSE}
# save
path <- paste0("../results/feature/",treatment,
               "_UMAP_unannotated_nCount")
f1
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)

# save
path <- paste0("../results/feature/",treatment,
               "_UMAP_unannotated_nFeature")
f2
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)


# save
path <- paste0("../results/feature/",treatment,
               "_UMAP_unannotated_percentMT")
f3
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)


# save
path <- paste0("../results/feature/",treatment,
               "_UMAP_unannotated_complexity")
f4
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)


remove(f1,f2,f3,f4)
```


## Percent cells per cluster
```{r percent_cells_per_cluster}
dataObject.unannotated$seurat_clusters <- dataObject.unannotated$SCT_snn_res.0.4

# sample
b2 <- dataObject.unannotated@meta.data %>%
  group_by(seurat_clusters, sample) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=sample)) +
  geom_col() +
  scale_fill_manual(values = sample_colors) +
  ggtitle("Percentage of sample per cluster")
b2
```

```{r,echo=FALSE}
# save
path <- paste0("../results/nuclei_count/",treatment,
               "_percent_nuclei_per_cluster_sample")
b2
saveToPDF(paste0(path, ".pdf"), width = 8, height = 4)

```


## Number cells per cluster
```{r cells_per_cluster}
treatment_ncells <- FetchData(dataObject.unannotated, 
                     vars = c("ident", "group")) %>%
  dplyr::count(ident, group) %>%
  tidyr::spread(ident, n)
write.table(treatment_ncells, 
            paste0("../results/nuclei_count/",
                   treatment, 
                   "_cells_per_cluster_treatment.txt"),
            quote = FALSE, sep = "\t")

sample_ncells <- FetchData(dataObject.unannotated, 
                     vars = c("ident", "sample")) %>%
  dplyr::count(ident,sample) %>%
  tidyr::spread(ident, n)
write.table(sample_ncells, 
            paste0("../results/nuclei_count/",
                   treatment, 
                   "_cells_per_cluster_sample.txt"),
            quote = FALSE, sep = "\t")

sample_ncells
```
# Save
```{r save_object_integrated,echo=FALSE,eval=TRUE}
# Save integrated seurat object
#DefaultAssay(dataObject.unannotated) <- "RNA"
#dataObject.unannotated$seurat_clusters <- dataObject.unannotated$SCT_snn_res.0.4
saveRDS(dataObject.unannotated, paste0("../rObjects/",treatment,"_unannotated_no_integration.rds"))
```


## Integration 
```{r integrate}
dataObject.integrated <- IntegrateLayers(
  object = dataObject.filtered, method = HarmonyIntegration,
  orig.reduction = "pca", new.reduction = "harmony",
  verbose = FALSE
)
dataObject.integrated <- RunUMAP(dataObject.integrated, reduction = "harmony", dims = 1:min.pc, reduction.name = "umap.harmony")

# Determine the K-nearest neighbor graph
dataObject.integrated <- FindNeighbors(object = dataObject.integrated, 
                                 reduction = "harmony", # pca, harmony 
                                 dims = 1:min.pc)

# Determine the clusters for various resolutions
dataObject.integrated <- FindClusters(object = dataObject.integrated,
                                 algorithm = 1, # 1= Louvain
                                 resolution = seq(0.1,1.2,by=0.1))

p1 <- DimPlot(
  dataObject.integrated,
  reduction = "umap.harmony",
  group.by = c("sample"),
  combine = FALSE, label.size = 2
)
p1

dataObject.integrated
```

## Explore resolutions
```{r}
# 0.1
umap0.1 <- DimPlot(dataObject.integrated,
        group.by = "SCT_snn_res.0.1",
        label = TRUE)
umap0.1

# 0.4
umap0.4 <- DimPlot(dataObject.integrated,
        group.by = "SCT_snn_res.0.4",
        label = TRUE)
umap0.4

# 0.6
umap0.6 <- DimPlot(dataObject.integrated,
        group.by = "SCT_snn_res.0.6",
        label = TRUE)
umap0.6

# 0.8
umap0.8 <- DimPlot(dataObject.integrated,
        group.by = "SCT_snn_res.0.8",
        label = TRUE)
umap0.8

```

```{r,echo=FALSE,message=FALSE,warning=FALSE}
# save umap0.1
path <- paste0("../results/UMAP/unannotated/",treatment,
               "_integrated_UMAP_unannotated_res0.1")
umap0.1
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# save umap0.4
path <- paste0("../results/UMAP/unannotated/",treatment,
               "_integrated_UMAP_unannotated_res0.4")
umap0.4
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# save umap0.6
path <- paste0("../results/UMAP/unannotated/",treatment,
               "_integrated_UMAP_unannotated_res0.6")
umap0.6
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)


# save umap0.8
path <- paste0("../results/UMAP/unannotated/",treatment,
               "_integrated_UMAP_unannotated_res0.8")
umap0.8
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)
```

## Clustering QC
## sample, type
```{r umap_treatment}
# sample
u1 <- DimPlot(dataObject.integrated, 
        label = FALSE,
        split.by = "sample") +
  NoLegend()
u1

# sample
u2 <- DimPlot(dataObject.integrated, 
        label = FALSE,
        split.by = "TYPE") +
  NoLegend()
u2

# save 
path <- paste0("../results/UMAP/unannotated/",treatment,
               "_SAMPLE_integrated_UMAP")
u1
saveToPDF(paste0(path, ".pdf"), width = 10, height = 5)

path <- paste0("../results/UMAP/unannotated/",treatment,
               "_TYPE_integrated_UMAP")
u2
saveToPDF(paste0(path, ".pdf"), width = 8, height = 5)
```


## Revisit QC metrics
```{r umap_QC}
# nCount
f1 <- FeaturePlot(dataObject.integrated, 
            features = "nCount_RNA",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f1

# nFeature
f2 <- FeaturePlot(dataObject.integrated, 
            features = "nFeature_RNA",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f2

# percent.mt
f3 <- FeaturePlot(dataObject.integrated, 
            features = "percent.mt",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f3

# cell.complexity
f4 <- FeaturePlot(dataObject.integrated, 
            features = "cell.complexity",
            pt.size = 0.4, 
            order = TRUE,
            min.cutoff = 'q10',
            label = TRUE)
f4

```

```{r,echo=FALSE,message=FALSE,warning=FALSE}
# save
path <- paste0("../results/feature/",treatment,
               "_integrated_UMAP_unannotated_nCount")
f1
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)

# save
path <- paste0("../results/feature/",treatment,
               "_integrated_UMAP_unannotated_nFeature_integrated")
f2
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)


# save
path <- paste0("../results/feature/",treatment,
               "_integrated_UMAP_unannotated_percentMT")
f3
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)


# save
path <- paste0("../results/feature/",treatment,
               "_integrated_UMAP_unannotated_complexity")
f4
saveToPDF(paste0(path, ".pdf"), width = 5.5, height = 4)


remove(f1,f2,f3,f4)
```


## Percent cells per cluster
```{r percent_cells_per_cluster}
dataObject.integrated$seurat_clusters <- dataObject.integrated$SCT_snn_res.0.4

# sample
b2 <- dataObject.integrated@meta.data %>%
  group_by(seurat_clusters, sample) %>%
  dplyr::count() %>%
  group_by(seurat_clusters) %>%
  dplyr::mutate(percent = 100*n/sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=seurat_clusters,y=percent, fill=sample)) +
  geom_col() +
  scale_fill_manual(values = sample_colors) +
  ggtitle("Percentage of sample per cluster")
b2
```

```{r,echo=FALSE}
# save
path <- paste0("../results/nuclei_count/",treatment,
               "_integrated_percent_nuclei_per_cluster_sample")
b2
saveToPDF(paste0(path, ".pdf"), width = 8, height = 4)

```


## Number cells per cluster
```{r cells_per_cluster}
treatment_ncells <- FetchData(dataObject.integrated, 
                     vars = c("ident", "TYPE")) %>%
  dplyr::count(ident, TYPE) %>%
  tidyr::spread(ident, n)
write.table(treatment_ncells, 
            paste0("../results/nuclei_count/",
                   treatment, 
                   "_integrated_cells_per_cluster_treatment.txt"),
            quote = FALSE, sep = "\t")

sample_ncells <- FetchData(dataObject.integrated, 
                     vars = c("ident", "sample")) %>%
  dplyr::count(ident,sample) %>%
  tidyr::spread(ident, n)
write.table(sample_ncells, 
            paste0("../results/nuclei_count/",
                   treatment, 
                   "_integrated_cells_per_cluster_sample.txt"),
            quote = FALSE, sep = "\t")

sample_ncells
```
# Save
```{r save_object_integrated,echo=FALSE,eval=TRUE}
# Save integrated seurat object
DefaultAssay(dataObject.integrated) <- "RNA"
dataObject.integrated$seurat_clusters <- dataObject.integrated$SCT_snn_res.0.4
saveRDS(dataObject.integrated, paste0("../rObjects/",treatment,"_unannotated_integrated.rds"))
```

```{r}
```

